### PR TITLE
Remove p-cancelable and refactor TokenProvidingSubscription

### PR DESCRIPTION
**Pull request by hamchapman**
### What?

- Remove p-cancelable 
- Refactor TokenProvidingSubscription

### Why?

p-cancelable seemed an unnecessary dependency and made some of the code and interfaces we exposed a bit awkward.

### How?

Refactor

----

CC @pusher/sigsdk

### CODE REVIEW

File: src/token-providing-subscription.ts [hunk 1]
```diff
@@ -1,147 +1,161 @@
...
+      new TokenProvidingSubscription(
+        logger,
+        listeners,
+        headers,
+        tokenProvider,
+        nextSubscribeStrategy
+      );
+  } else {
+    return (listeners, headers) =>
+      nextSubscribeStrategy(listeners, headers);
```
Comment (pl) [inline src/token-providing-subscription.ts]: You can just return `nextSubscribeStrategy` without wrapping it in an extra function.

File: src/token-providing-subscription.ts [hunk 2]
```diff
@@ -1,147 +1,161 @@
...
+  }
+}
 
-            class OpenSubscriptionState implements SubscriptionState {
-                constructor(private headers: ElementsHeaders, private underlyingSubscription: Subscription, private onTransition: (SubscriptionState) => void){
-                    logger.verbose(`TokenProvidingSubscription: transitioning to OpenSubscriptionState`);
-                    listeners.onOpen(headers);
-                }
+export interface TokenProvidingSubscriptionState {
```
Comment (pl) [inline src/token-providing-subscription.ts]: This type should not be exported.

File: src/token-providing-subscription.ts [hunk 3]
```diff
@@ -1,147 +1,161 @@
...
+  unsubscribe(): void;
+}
 
-                unsubscribe(){
-                    this.underlyingSubscription.unsubscribe();
-                    this.onTransition(new EndedSubscriptionState());
-                }
+class TokenProvidingSubscription implements Subscription {
+  private state: TokenProvidingSubscriptionState;
+  private tokenPromise: Promise<any>;
```
Comment (pl) [inline src/token-providing-subscription.ts]: You don't need `tokenPromise` to be a property. It's only used in `subscribe` method scope.

File: src/token-providing-subscription.ts [hunk 4]
```diff
@@ -1,147 +1,161 @@
...
+
+  private isTokenExpiredError(error: any): boolean {
+    return (
+      error instanceof ErrorResponse &&
+      error.statusCode === 401 &&
+      error.info === "authentication/expired"
+    );
+  }
+
+  private onTransition = (newState: TokenProvidingSubscriptionState) => {
```
Comment (pl) [inline src/token-providing-subscription.ts]: This method is a bit excessive :P

File: src/token-providing-subscription.ts [hunk 5]
```diff
@@ -1,147 +1,161 @@
...
+                this.tokenProvider.clearToken(token);
+                this.subscribe();
+              } else {
+                this.onTransition(new InactiveState(this.logger));
+                existingListeners.onError(error);
+              }
             }
+          }
+        );
+        this.state.subscribe(token, listenersWithWrappedOnError);
```
Comment (pl) [inline src/token-providing-subscription.ts]: I think constructing the listeners manually is more direct than using `Object.assign`:

```
this.state.subscribe(token, {
  onOpen: this.listeners.onOpen,
  onEvent: this.listeners.onEvent,
  onEnd: (error: any) => {
    this.onTransition(new InactiveState(this.logger));
    existingListeners.onEnd(error);
  },
  onError: (error: any) => {
    if (this.isTokenExpiredError(error)) {
      this.tokenProvider.clearToken(token);
      this.subscribe();
    } else {
      this.onTransition(new InactiveState(this.logger));
      existingListeners.onError(error);
    }
  }
});
```

I'm a bit surprised that it compiled without warnings. Maybe types are not strict enough.

File: src/token-providing-subscription.ts [hunk 6]
```diff
@@ -1,147 +1,161 @@
...
+          listeners.onEnd(error);
+        }
+      },
+      this.headers
+    )
+  }
+
+  unsubscribe() {
+    this.underlyingSubscription.unsubscribe();
+    this.onTransition(new InactiveState(this.logger));
```
Comment (pl) [inline src/token-providing-subscription.ts]: This line should be removed.

**Comment by hamchapman** [comment]
Feedback addressed - thank you! Good to merge?

**Comment by hamchapman** [comment]
Likewise. We have some very basic ones but they're broken on master (but fixed in #75). #75 also fixes the integration tests, which are definitely more useful in checking for regressions.

I'll merge this and then rebase #75 on master and then we can beautifully formatted and linted code with working unit and integration tests.

**Pull request merged**